File: JasSwift/Tests/Interpreter/ToolOptionsDispatchTest.swift
/// Unit oracle for the toolbar tool-options double-click lookup +
/// dispatch resolution
/// (`Sources/Interpreter/WorkspaceLoader.swift` →
/// `resolveToolOptions` / `isToolButtonElement`).
///
/// Double-clicking a TOOLBAR tool button opens the ACTIVE tool's
/// options. The active tool's entry is looked up in the compiled bundle
/// `tools` map (NOT a hardcoded list) and its options are dispatched in
/// priority order: `tool_options_panel` → `tool_options_action` →
/// `tool_options_dialog`. A tool that declares none is a no-op.
///
/// The double-click is scoped to toolbar tool slots only: an
/// `icon_button` whose `behavior` carries a `click` event dispatching
/// `select_tool`. Panels and every other `icon_button` never match.

import Foundation
import Testing
@testable import JasLib

// MARK: - resolveToolOptions (priority + bundle-driven)

@Test func resolveToolOptionsReadsDialogFromBundle() {
    // The live bundle's paintbrush/blob_brush/scale/rotate/shear/
    // eyedropper tools declare tool_options_dialog. Resolve straight
    // from the compiled bundle (no hardcoded tool list).
    let tools = WorkspaceData.load()?.data["tools"] as? [String: Any] ?? [:]
    #expect(!tools.isEmpty, "bundle tools map should be present")

    #expect(resolveToolOptions(tools: tools, activeTool: "paintbrush")
            == .dialog("paintbrush_tool_options"))
    #expect(resolveToolOptions(tools: tools, activeTool: "blob_brush")
            == .dialog("blob_brush_tool_options"))
    #expect(resolveToolOptions(tools: tools, activeTool: "eyedropper")
            == .dialog("eyedropper_tool_options"))
    #expect(resolveToolOptions(tools: tools, activeTool: "scale")
            == .dialog("scale_options"))
    #expect(resolveToolOptions(tools: tools, activeTool: "rotate")
            == .dialog("rotate_options"))
    #expect(resolveToolOptions(tools: tools, activeTool: "shear")
            == .dialog("shear_options"))
}

@Test func resolveToolOptionsReadsPanelFromBundle() {
    // magic_wand declares tool_options_panel → magic_wand (the only
    // panel-options tool in the bundle today).
    let tools = WorkspaceData.load()?.data["tools"] as? [String: Any] ?? [:]
    #expect(resolveToolOptions(tools: tools, activeTool: "magic_wand")
            == .panel("magic_wand"))
    // …and the panel id maps to a real PanelKind, so the dispatch can
    // actually show a panel.
    #expect(panelIdToKind("magic_wand") == .magicWand)
}

@Test func resolveToolOptionsReadsActionFromBundle() {
    // hand → fit_active_artboard, zoom → zoom_to_actual_size,
    // artboard → fit_all_artboards. These are the view-action arm the
    // old native path never dispatched.
    let tools = WorkspaceData.load()?.data["tools"] as? [String: Any] ?? [:]
    #expect(resolveToolOptions(tools: tools, activeTool: "hand")
            == .action("fit_active_artboard"))
    #expect(resolveToolOptions(tools: tools, activeTool: "zoom")
            == .action("zoom_to_actual_size"))
    #expect(resolveToolOptions(tools: tools, activeTool: "artboard")
            == .action("fit_all_artboards"))
}

@Test func resolveToolOptionsNoneWhenToolDeclaresNoOptions() {
    // The plain Selection tool declares no options field → no-op.
    let tools = WorkspaceData.load()?.data["tools"] as? [String: Any] ?? [:]
    #expect(resolveToolOptions(tools: tools, activeTool: "selection") == ToolOptionsDispatch.none)
    // An unknown tool id also resolves to .none (no crash, no-op).
    #expect(resolveToolOptions(tools: tools, activeTool: "no_such_tool") == ToolOptionsDispatch.none)
}

@Test func resolveToolOptionsPriorityPanelBeatsActionBeatsDialog() {
    // Synthetic specs verifying the priority order independent of the
    // live bundle: panel wins over action wins over dialog.
    let allThree: [String: Any] = [
        "x": [
            "tool_options_panel": "magic_wand",
            "tool_options_action": "fit_active_artboard",
            "tool_options_dialog": "paintbrush_tool_options",
        ] as [String: Any]
    ]
    #expect(resolveToolOptions(tools: allThree, activeTool: "x") == .panel("magic_wand"))

    let actionAndDialog: [String: Any] = [
        "x": [
            "tool_options_action": "zoom_to_actual_size",
            "tool_options_dialog": "scale_options",
        ] as [String: Any]
    ]
    #expect(resolveToolOptions(tools: actionAndDialog, activeTool: "x")
            == .action("zoom_to_actual_size"))

    // Empty-string fields are treated as absent (skip to next).
    let emptyPanel: [String: Any] = [
        "x": [
            "tool_options_panel": "",
            "tool_options_dialog": "scale_options",
        ] as [String: Any]
    ]
    #expect(resolveToolOptions(tools: emptyPanel, activeTool: "x")
            == .dialog("scale_options"))
}

// MARK: - isToolButtonElement (dblclick scoping)

@Test func isToolButtonTrueForToolbarSlots() {
    // Every tool slot in the bundle's tool_grid carries a click →
    // select_tool behavior, so all of them are tool buttons.
    let grid = WorkspaceData.load()?.toolGrid()
    let children = grid?["children"] as? [[String: Any]] ?? []
    #expect(!children.isEmpty, "tool_grid should have icon_button children")
    for child in children {
        #expect(isToolButtonElement(child),
                "tool_grid child \(child["id"] ?? "?") should be a tool button")
    }
}

@Test func isToolButtonFalseForNonToolIconButtons() {
    // A panel radio / dialog glyph icon_button with no select_tool click
    // must NOT match — the dblclick stays off panels and other buttons.
    let panelRadio: [String: Any] = [
        "type": "icon_button",
        "icon": "align_left",
        "behavior": [
            ["event": "click", "action": "align", "params": ["edge": "left"]]
        ],
    ]
    #expect(!isToolButtonElement(panelRadio))

    // An icon_button with only press behaviors (e.g. a long-press-only
    // affordance) and no select_tool click also does not match.
    let pressOnly: [String: Any] = [
        "type": "icon_button",
        "behavior": [
            ["event": "mouse_down", "effects": []],
            ["event": "mouse_up", "effects": []],
        ],
    ]
    #expect(!isToolButtonElement(pressOnly))

    // A non-icon_button element (slider, text, …) never matches even if
    // it somehow carried a select_tool behavior.
    let slider: [String: Any] = [
        "type": "slider",
        "behavior": [["event": "click", "action": "select_tool"]],
    ]
    #expect(!isToolButtonElement(slider))

    // An icon_button with no behavior at all does not match.
    #expect(!isToolButtonElement(["type": "icon_button"]))
}

@Test func isToolButtonTrueForLongPressSlot() {
    // A long-press-alternate slot carries mouse_down / mouse_up AND a
    // click → select_tool. It IS a tool button (clicking commits the
    // visible tool), so it must respond to the options dblclick too —
    // matching the prior native toolButtonWithAlternates.
    let alternateSlot: [String: Any] = [
        "type": "icon_button",
        "behavior": [
            ["event": "mouse_down", "effects": []],
            ["event": "mouse_up", "effects": []],
            ["event": "click", "action": "select_tool", "params": ["tool": "pen"]],
        ],
    ]
    #expect(isToolButtonElement(alternateSlot))
}

File: jas_ocaml/test/interpreter/tool_options_dispatch_test.ml
open Jas

(* Unit tests for the toolbar tool-options double-click lookup/dispatch
   (Yaml_panel_view). These exercise the PURE, GUI-free surface:
     - [is_tool_button] (the discriminator that scopes the dblclick to
       toolbar tool buttons only), and
     - [tool_options_dispatch_for] (the bundle lookup + panel > action >
       dialog priority order).
   The actual panel/dialog opening + action run is GUI (user-verified). *)

module Y = Yaml_panel_view

(* Build a minimal icon_button element with the given behavior array. *)
let icon_button ~behavior =
  `Assoc [ ("type", `String "icon_button"); ("behavior", `List behavior) ]

let click_select_tool tool =
  `Assoc [ ("event", `String "click");
           ("action", `String "select_tool");
           ("params", `Assoc [ ("tool", `String tool) ]) ]

(* ── is_tool_button ──────────────────────────────────────────────── *)

let is_tool_button_tests = [
  Alcotest.test_case "tool_button_with_select_tool_click" `Quick (fun () ->
    let el = icon_button ~behavior:[ click_select_tool "pencil" ] in
    Alcotest.(check bool) "tool button" true (Y.is_tool_button el));

  Alcotest.test_case "slot_tool_button_with_long_press" `Quick (fun () ->
    (* A slot button has mouse_down/up flyout behaviors AND the click
       select_tool — still a tool button. *)
    let el = icon_button ~behavior:[
      `Assoc [ ("event", `String "mouse_down"); ("effects", `List []) ];
      `Assoc [ ("event", `String "mouse_up"); ("effects", `List []) ];
      click_select_tool "pen";
    ] in
    Alcotest.(check bool) "slot tool button" true (Y.is_tool_button el));

  Alcotest.test_case "fill_stroke_button_is_not_a_tool_button" `Quick (fun () ->
    (* The swap / reset / solid / gradient / none icon_buttons use other
       actions, not select_tool — must be excluded. *)
    let el = icon_button ~behavior:[
      `Assoc [ ("event", `String "click");
               ("action", `String "swap_fill_stroke") ];
    ] in
    Alcotest.(check bool) "fill/stroke button excluded"
      false (Y.is_tool_button el));

  Alcotest.test_case "panel_icon_button_without_behavior_excluded" `Quick (fun () ->
    let el = `Assoc [ ("type", `String "icon_button") ] in
    Alcotest.(check bool) "no behavior excluded" false (Y.is_tool_button el));

  Alcotest.test_case "plain_button_type_excluded" `Quick (fun () ->
    (* Even with a select_tool click, a non-icon_button type isn't a tool
       button (the toolbar grid stamps icon_button). *)
    let el = `Assoc [ ("type", `String "button");
                      ("behavior", `List [ click_select_tool "pencil" ]) ] in
    Alcotest.(check bool) "plain button excluded" false (Y.is_tool_button el));
]

(* ── tool_options_dispatch_for (bundle lookup + priority) ────────── *)

(* These read the compiled workspace.json, so they only run when the
   bundle loads. Guard each by checking the bundle is present; if not,
   skip the assertion (the cross-language corpus gate covers bundle
   presence elsewhere). The expected mappings come straight from the
   bundle's tools map:
     paintbrush/blob_brush/scale/rotate/shear/eyedropper -> dialog
     hand/zoom/artboard -> action
     magic_wand -> panel
     pencil/selection -> none. *)
let bundle_loaded () = Workspace_loader.load () <> None

let dispatch_tests = [
  Alcotest.test_case "magic_wand_opens_panel" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "magic_wand" with
      | Y.Show_panel Workspace_layout.Magic_wand -> ()
      | _ -> Alcotest.fail "magic_wand should resolve to Show_panel Magic_wand");

  Alcotest.test_case "hand_runs_action" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "hand" with
      | Y.Run_action "fit_active_artboard" -> ()
      | _ -> Alcotest.fail "hand should resolve to Run_action fit_active_artboard");

  Alcotest.test_case "zoom_runs_action" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "zoom" with
      | Y.Run_action "zoom_to_actual_size" -> ()
      | _ -> Alcotest.fail "zoom should resolve to Run_action zoom_to_actual_size");

  Alcotest.test_case "paintbrush_opens_dialog" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "paintbrush" with
      | Y.Open_dialog "paintbrush_tool_options" -> ()
      | _ -> Alcotest.fail "paintbrush should resolve to Open_dialog");

  Alcotest.test_case "scale_opens_dialog" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "scale" with
      | Y.Open_dialog "scale_options" -> ()
      | _ -> Alcotest.fail "scale should resolve to Open_dialog scale_options");

  Alcotest.test_case "eyedropper_opens_dialog" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "eyedropper" with
      | Y.Open_dialog "eyedropper_tool_options" -> ()
      | _ -> Alcotest.fail "eyedropper should resolve to Open_dialog");

  Alcotest.test_case "pencil_has_no_options" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "pencil" with
      | Y.No_options -> ()
      | _ -> Alcotest.fail "pencil declares no tool options -> No_options");

  Alcotest.test_case "selection_has_no_options" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "selection" with
      | Y.No_options -> ()
      | _ -> Alcotest.fail "selection declares no tool options -> No_options");

  Alcotest.test_case "unknown_tool_has_no_options" `Quick (fun () ->
    if bundle_loaded () then
      match Y.tool_options_dispatch_for "no_such_tool_xyz" with
      | Y.No_options -> ()
      | _ -> Alcotest.fail "unknown tool -> No_options");
]

(* ── panel_id_to_kind ────────────────────────────────────────────── *)

let panel_id_tests = [
  Alcotest.test_case "magic_wand_id_maps" `Quick (fun () ->
    Alcotest.(check bool) "magic_wand -> Magic_wand" true
      (Y.panel_id_to_kind "magic_wand" = Some Workspace_layout.Magic_wand));

  Alcotest.test_case "color_id_maps" `Quick (fun () ->
    Alcotest.(check bool) "color -> Color" true
      (Y.panel_id_to_kind "color" = Some Workspace_layout.Color));

  Alcotest.test_case "unknown_panel_id_is_none" `Quick (fun () ->
    Alcotest.(check bool) "unknown -> None" true
      (Y.panel_id_to_kind "no_such_panel" = None));
]

let () =
  Alcotest.run "tool_options_dispatch" [
    "is_tool_button", is_tool_button_tests;
    "dispatch_for", dispatch_tests;
    "panel_id_to_kind", panel_id_tests;
  ]
